src/engine/input/key-mapping/InputEventItem.cs

```diff
@@ -0,0 +1,138 @@
+    public NodePath ButtonPath;
+
+    internal InputActionItem AssociatedAction;
+
+    internal bool JustAdded;
+
+    private Button button;
+
+    private bool waitingForInput;
+
+    public InputEventWithModifiers AssociatedEvent { get; set; }
+
+    public void Delete()
+    {
+        AssociatedAction.AssociatedGroup.GetParent<InputGroupList>().ControlSchemeChanged();
+        AssociatedAction.Inputs.Remove(this);
+    }
+
+    public override void _Input(InputEvent @event)
+    {
+        if (InputGroupList.Instance.IsConflictDialogOpen())
+            return;
+
+        if (!waitingForInput)
+            return;
+
+        if (!(@event is InputEventMouseButton) && !(@event is InputEventKey))
+            return;
+
+        if (@event is InputEventKey iek)
+        {
+            switch (iek.Scancode)
+            {
+                case (uint)KeyList.Escape:
+                    Delete();
+                    return;
+                case (uint)KeyList.Alt:
+                case (uint)KeyList.Shift:
+                case (uint)KeyList.Control:
+                    return;
+            }
+        }
+
+        var old = AssociatedEvent;
+        AssociatedEvent = (InputEventWithModifiers)@event;
+        var conflict = InputGroupList.Instance.Conflicts(this);
+        if (conflict != null)
+        {
+            InputGroupList.Instance.ShowDialog(this, conflict, AssociatedEvent);
+            AssociatedEvent = old;
+            return;
+        }
+
+        waitingForInput = false;
+        JustAdded = false;
+        AssociatedAction.AssociatedGroup.GetParent<InputGroupList>().ControlSchemeChanged();
+
+        UpdateButtonText();
+    }
+
+    public override void _Ready()
+    {
+        button = GetNode<Button>(ButtonPath);
+        if (JustAdded)
+            OnButtonPressed();
+        else
+            UpdateButtonText();
+    }
+
+    public override bool Equals(object obj)
+    {
+        if (!(obj is InputEventItem input))
+            return false;
+
+        return string.Equals(AsText(), input.AsText(), StringComparison.InvariantCulture);
+    }
+
+    public override int GetHashCode()
+    {
+        return AsText().GetHashCode();
+    }
+
+    private void OnButtonPressed()
+    {
+        waitingForInput = true;
+        button.Text = "Press a key...";
+    }
+
+    private void UpdateButtonText()
+    {
+        button.Text = AsText();
+    }
+
+    private string AsText()
+    {
+        var text = string.Empty;
+        if (AssociatedEvent == null)
+            return text;
+
+        switch (AssociatedEvent)
+        {
+            case InputEventMouseButton iemb:
+                if (AssociatedEvent.Control)
+                    text += "Control+";
+                if (AssociatedEvent.Alt)
+                    text += "Alt+";
+                if (AssociatedEvent.Shift)
+                    text += "Shift+";
+                text += iemb.ButtonIndex switch
+                {
+                    1 => "Left mouse",
```

Comment:
Would it be possible to get these strings from godot?

Reply:
no, because the AsText and the ToString methods of the InputMouseButton return a technical string